<a href="https://colab.research.google.com/github/wcj365/python-stats-dataviz/blob/master/world_development_indicator.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Chapter 4 - World Development Indicator

**Data Source**
- https://datacatalog.worldbank.org/dataset/world-development-indicators
- https://unstats.un.org/unsd/methodology/m49/overview

In [1]:
!wget http://databank.worldbank.org/data/download/WDI_csv.zip
!unzip WDI_csv.zip

--2021-01-20 00:47:55--  http://databank.worldbank.org/data/download/WDI_csv.zip
Resolving databank.worldbank.org (databank.worldbank.org)... 192.86.98.102
Connecting to databank.worldbank.org (databank.worldbank.org)|192.86.98.102|:80... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://databank.worldbank.org/data/download/WDI_csv.zip [following]
--2021-01-20 00:47:55--  https://databank.worldbank.org/data/download/WDI_csv.zip
Connecting to databank.worldbank.org (databank.worldbank.org)|192.86.98.102|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 66502732 (63M) [application/x-zip-compressed]
Saving to: ‘WDI_csv.zip’

WDI_csv.zip         100%[===================>]  63.42M   378KB/s    in 69s     

2021-01-20 00:49:04 (935 KB/s) - ‘WDI_csv.zip’ saved [66502732/66502732]



In [3]:
import pandas as pd
import plotly.express as px
import plotly.graph_objs as go
import plotly.io as pio

## Step 1 - Read Data

In [4]:
df = pd.read_csv("WDIData.csv")
df.shape

(380160, 66)

In [5]:
df.sample(2)

Country Name Country Code  ... 2020 Unnamed: 65
165140             Finland          FIN  ...  NaN         NaN
50452   Other small states          OSS  ...  NaN         NaN

[2 rows x 66 columns]

**Drop the last unuseful column**

In [6]:
df.drop(columns=df.columns[-1], inplace=True)
df.sample(1)

Country Name Country Code  ... 2019 2020
298896      Romania          ROU  ...  NaN  NaN

[1 rows x 65 columns]

## Step 2 - Add additional variables for regions and sub-regions

In [8]:
df_country = pd.read_csv("country_regions.csv")
df_country.head(2)

Global Code  ... Developed / Developing Countries
0            1  ...                       Developing
1            1  ...                       Developing

[2 rows x 16 columns]

In [9]:
df_country = df_country[["ISO-alpha3 Code", "Region Name", "Sub-region Name", "Country or Area"]]
df_country.sample(2)

ISO-alpha3 Code  ... Country or Area
153             CYP  ...          Cyprus
110             URY  ...         Uruguay

[2 rows x 4 columns]

In [10]:
df_merged= pd.merge(df_country, 
                    df,
                    right_on=["Country Code"],
                    left_on=["ISO-alpha3 Code"],
                    how="right"
)

df_merged.sample(5)

ISO-alpha3 Code Region Name  ...          2019 2020
346097             TLS        Asia  ...  1.502372e+03  NaN
372071             VNM        Asia  ...  1.782288e+11  NaN
334514             SDN      Africa  ...           NaN  NaN
108532             BRN        Asia  ...  5.507711e+09  NaN
279745             NOR      Europe  ...           NaN  NaN

[5 rows x 69 columns]

## Step 3 - Melt the dataframe for easy analysis

In [11]:
df_melted = df_merged.melt(id_vars=df_merged.columns[:8], 
                          value_vars=df_merged.columns[8:], 
                          var_name="Year", 
                          value_name='Indicator Value')

df_melted.sample(5)

ISO-alpha3 Code Region Name  ...  Year Indicator Value
20836655             SAU        Asia  ...  2014            3.84
11320178             PRI    Americas  ...  1989             NaN
20905570             YEM        Asia  ...  2014           83.70
22870332             NaN         NaN  ...  2020             NaN
22156384             VGB    Americas  ...  2018             NaN

[5 rows x 10 columns]

**Find out how many unique indicators**

In [12]:
df["Indicator Code"].nunique()

1440

**1,440 indicators are a lot. Save them to a file for reference**

In [13]:
df[["Indicator Name", "Indicator Code"]].drop_duplicates().to_csv("indicators.csv", index=False)

**Eye ball the file and pick the three indicators of interest:**
- SP.DYN.LE00.IN
    - Life expectancy at birth, total (years)
- NY.GDP.PCAP.PP.CD
    - GDP per capita, PPP (current international $)
- SP.POP.TOTL
    - Population, total 

In [14]:
indicator_list = ["SP.DYN.LE00.IN","NY.GDP.PCAP.PP.CD","SP.POP.TOTL"]
df_plot = df_melted[df_melted["Indicator Code"].isin(indicator_list)]
df_plot.shape

(48312, 10)

In [15]:
df_plot.sample(5)

ISO-alpha3 Code Region Name  ...  Year Indicator Value
4425597              MLI      Africa  ...  1971             NaN
4048540              MHL     Oceania  ...  1970             NaN
21658300             VUT     Oceania  ...  2016          70.021
10040380             ERI      Africa  ...  1986          48.230
23115943             STP      Africa  ...  2020             NaN

[5 rows x 10 columns]

In [16]:
indicator_columns = ("Indicator Name", "Indicator Code", "Indicator Value")
index_columns = [column for column in df_plot.columns if column not in indicator_columns]
index_columns

['ISO-alpha3 Code',
 'Region Name',
 'Sub-region Name',
 'Country or Area',
 'Country Name',
 'Country Code',
 'Year']

In [17]:
df_plot = df_plot.pivot(index=index_columns, columns=["Indicator Code"], values="Indicator Value")
df_plot.sample(5)

Indicator Code                                                                                      NY.GDP.PCAP.PP.CD  ...  SP.POP.TOTL
ISO-alpha3 Code Region Name Sub-region Name Country or Area Country Name         Country Code Year                     ...             
ALB             Europe      Southern Europe Albania         Albania              ALB          2005        5865.324803  ...    3011487.0
NaN             NaN         NaN             NaN             Hong Kong SAR, China HKG          2017       59849.248176  ...    7391700.0
BEL             Europe      Western Europe  Belgium         Belgium              BEL          1967                NaN  ...    9580991.0
MCO             Europe      Western Europe  Monaco          Monaco               MCO          1968                NaN  ...      23304.0
KGZ             Asia        Central Asia    Kyrgyzstan      Kyrgyz Republic      KGZ          1965                NaN  ...    2573300.0

[5 rows x 3 columns]

In [18]:
df_plot.reset_index(inplace=True)
df_plot.sample(5)

Indicator Code ISO-alpha3 Code Region Name  ... SP.DYN.LE00.IN  SP.POP.TOTL
7296                       GEO        Asia  ...      69.613000    4349913.0
2023                       NaN         NaN  ...      52.149211  127335091.0
7076                       FSM     Oceania  ...      54.513000      44514.0
9450                       KOR        Asia  ...      82.275610   51217803.0
14323                      SYR        Asia  ...      72.938000   21205873.0

[5 rows x 10 columns]

In [19]:
df_plot = df_plot[df_plot["Sub-region Name"].notnull()]
df_plot = df_plot[df_plot["Region Name"].notnull()]
df_plot = df_plot[df_plot["SP.POP.TOTL"].notnull()]

In [20]:
df_plot.sample(5)

Indicator Code ISO-alpha3 Code Region Name  ... SP.DYN.LE00.IN  SP.POP.TOTL
9069                       KAZ        Asia  ...      65.768293   14858335.0
3911                       BDI      Africa  ...      43.332000    3253218.0
4204                       BGD        Asia  ...      71.785000  157970840.0
15351                      USA    Americas  ...      76.636585  282162411.0
11352                      MYS        Asia  ...      62.833000    9790084.0

[5 rows x 10 columns]

In [21]:
df_plot.query("Year=='1989'").sample(10)

Indicator Code ISO-alpha3 Code Region Name  ... SP.DYN.LE00.IN SP.POP.TOTL
10460                      MDG      Africa  ...      50.530000  11268658.0
13205                      SDN      Africa  ...      55.334000  19476647.0
7288                       GEO        Asia  ...      70.425000   4803300.0
10216                      MAF    Americas  ...      74.321951     28722.0
14303                      SYR        Asia  ...      70.207000  12080444.0
15462                      VCT    Americas  ...      70.451000    107071.0
6007                       CYP        Asia  ...      76.394000    751047.0
3384                       AND      Europe  ...            NaN     52448.0
6251                       DMA    Americas  ...            NaN     70723.0
8386                       IMN      Europe  ...            NaN     69267.0

[10 rows x 10 columns]

In [22]:
df_scatter = df_plot.query("Year>'1989'").query("Year<'2019'")
df_scatter["Year"].unique()

array(['1990', '1991', '1992', '1993', '1994', '1995', '1996', '1997',
       '1998', '1999', '2000', '2001', '2002', '2003', '2004', '2005',
       '2006', '2007', '2008', '2009', '2010', '2011', '2012', '2013',
       '2014', '2015', '2016', '2017', '2018'], dtype=object)

In [23]:
df_scatter["NY.GDP.PCAP.PP.CD"].min()

285.586899959856

In [24]:
df_scatter["NY.GDP.PCAP.PP.CD"].max()

141634.96180221302

In [25]:
df_scatter[df_scatter["NY.GDP.PCAP.PP.CD"] > 140000]

Indicator Code ISO-alpha3 Code Region Name  ... SP.DYN.LE00.IN SP.POP.TOTL
12923                      QAT        Asia  ...         79.399   2196074.0

[1 rows x 10 columns]

In [26]:
df_scatter["SP.DYN.LE00.IN"].min()

26.171999999999997

In [29]:
fig = px.scatter(df_scatter, 
                 x="NY.GDP.PCAP.PP.CD", 
                 y="SP.DYN.LE00.IN",
          #       width=1000, 
         #        height=600,
                 color="Sub-region Name",
         #        color="Region Name",
                 size="SP.POP.TOTL",
          #       log_x=True, 
                 size_max=100,
                 hover_name="Country Name",
                 animation_group="Country Code",
                 animation_frame="Year",
                 range_x=[250,150000], 
                 range_y=[25,100]
              #  hover_data=["Country Code", "Country Name", "Sub-region Name"]
)

fig.write_html("scatter.html")
fig.show()

In [28]:
fig = px.choropleth(df_plot.query("Year=='2018'"),  
                    locations='Country Code', 
                    color='SP.POP.TOTL',
                    color_continuous_scale="Viridis",
                    scope="world",
                    hover_name="Country Name"
#                    locationmode = 'USA-states',
#                    labels={'ST':'State'}
)

#fig.update_layout(margin={"r":0,"t":0,"l":0,"b":0})
fig.write_html("choropleth.html")
fig.show()